internal/analytics/segment_reporter.go

```diff
@@ -158,12 +181,61 @@ func (r *SegmentReporter) ReportFatalError(err error) error {
 	return nil
 }
 
+// Run runs the reporter.
+func (r *SegmentReporter) Run(ctx context.Context) error {
+	r.log.Debug("Running heartbeat reporting...")
+
+	ticker := time.NewTicker(r.tickDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			err := r.reportHeartbeatEvent()
+			if err != nil {
+				return fmt.Errorf("while reporting heartbeat event: %w", err)
+			}
+
+			return nil
+		case <-ticker.C:
+			err := r.reportHeartbeatEvent()
+			if err != nil {
+				wrappedErr := fmt.Errorf("while reporting heartbeat event: %w", err)
+				r.log.Error(wrappedErr.Error())
```

Comment:
Personally, I like that approach more when it comes to logging, but it's up to you.
```suggestion
				r.log.WithError(err).Error("Failed to report heartbeat event")
```